### DISCUSSION TITLE
How can I use zustand persist with indexeddb?

### DISCUSSION THREAD

**Discussion by CherphAlt** [pmndrs/zustand#1721] [Q&A] [answered]
Indexeddb can store more data than localstorage and other types of storage ? How can I use it with zustand?
I saw that it can be used with indexeddb in the persist guide but there was no example

**Comment by dai-shi** [comment]
I think someone did persist+indexedeb before (but maybe with old options).
Hope someone can help on this.

**Comment by JacobWeisenburger** [comment]
feel free to use this template:
[Zustand custom persist storage template](https://gist.github.com/JacobWeisenburger/eabdb5cadc947e9d35f45d6b7cf2710e)

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Comment by fredericrous** [comment]
From the docs, the recommended way to use indexeddb is as follow

```ts
import { create } from 'zustand'
import { persist, createJSONStorage, StateStorage } from 'zustand/middleware'
import { get, set, del } from 'idb-keyval' // can use anything: IndexedDB, Ionic Storage, etc.

// Custom storage object
const storage: StateStorage = {
  getItem: async (name: string): Promise<string | null> => {
    console.log(name, 'has been retrieved')
    return (await get(name)) || null
  },
  setItem: async (name: string, value: string): Promise<void> => {
    console.log(name, 'with value', value, 'has been saved')
    await set(name, value)
  },
  removeItem: async (name: string): Promise<void> => {
    console.log(name, 'has been deleted')
    await del(name)
  },
}

export const useBoundStore = create(
  persist(
    (set, get) => ({
      bears: 0,
      addABear: () => set({ bears: get().bears + 1 }),
    }),
    {
      name: 'food-storage', // unique name
      storage: createJSONStorage(() => storage),
    },
  ),
)
```

https://docs.pmnd.rs/zustand/integrations/persisting-store-data#how-can-i-use-a-custom-storage-engine

**Comment by KingXP-Pythoner** [comment]
> From the docs, the recommended way to use indexeddb is as follow
> 
> ```ts
> import { create } from 'zustand'
> import { persist, createJSONStorage, StateStorage } from 'zustand/middleware'
> import { get, set, del } from 'idb-keyval' // can use anything: IndexedDB, Ionic Storage, etc.
> 
> // Custom storage object
> const storage: StateStorage = {
>   getItem: async (name: string): Promise<string | null> => {
>     console.log(name, 'has been retrieved')
>     return (await get(name)) || null
>   },
>   setItem: async (name: string, value: string): Promise<void> => {
>     console.log(name, 'with value', value, 'has been saved')
>     await set(name, value)
>   },
>   removeItem: async (name: string): Promise<void> => {
>     console.log(name, 'has been deleted')
>     await del(name)
>   },
> }
> 
> export const useBoundStore = create(
>   persist(
>     (set, get) => ({
>       bears: 0,
>       addABear: () => set({ bears: get().bears + 1 }),
>     }),
>     {
>       name: 'food-storage', // unique name
>       storage: createJSONStorage(() => storage),
>     },
>   ),
> )
> ```
> 
> https://docs.pmnd.rs/zustand/integrations/persisting-store-data#how-can-i-use-a-custom-storage-engine

I attempted to follow your answer and also referenced the docs in my Nextjs 14 stack project where I use Dexie.js instead of idb-keyval. However, im unable to rehydrate the store. I always get the warning that it couldn't find the storage.

**Reply by fredericrous** [reply, marked as answer]
with Remix I insure indexeddb is opened by putting in my [clientLoader](https://remix.run/docs/en/main/route/client-loader) a code like this

```ts
  const dbExists = await new Promise(resolve => {
    const request = indexedDB.open('app-storage')
    request.onsuccess = event => {
      resolve(true)
    }
    request.onerror = event => {
      resolve(false)
    }
  })
  const apps = dbExists ? selectApps(useAppsStore.getState()) : []
```

**Reply by KingXP-Pythoner** [reply]
Literally after you sent your reply, I realized I initialized my custom storage below the useStore and used createJSONStorage. For some reason, it didn't throw an error regarding using the object before initialization. Can't believe I made such a noob mistake. But thank you for your reply. It triggered my brain to realize the mistake haha.

**Comment by learnuidev** [comment]
Here is how I use index db and zustand:

```
import { create, StateCreator } from "zustand";
import { persist, createJSONStorage, StateStorage } from "zustand/middleware";
import { get, set, del } from "idb-keyval"; // can use anything: IndexedDB, Ionic Storage, etc.

// Custom storage object
export const indexDBStorage: StateStorage = {
  getItem: async (name: string): Promise<string | null> => {
    console.log(name, "has been retrieved");
    return (await get(name)) || null;
  },
  setItem: async (name: string, value: string): Promise<void> => {
    console.log(name, "with value", value, "has been saved");
    await set(name, value);
  },
  removeItem: async (name: string): Promise<void> => {
    console.log(name, "has been deleted");
    await del(name);
  },
};

export const createIndexDBStore = <T extends object>({
  name,
  handler,
}: {
  name: string;
  handler: StateCreator<T>;
}) =>
  create(
    persist(handler, {
      name, // name of the item in the storage (must be unique)
      storage: createJSONStorage(() => indexDBStorage), // (optional) by default, 'localStorage' is used
    })
  );

// Usage
export const useRepeatHistoryStore = createIndexDBStore({
  name: "mandarino/repeat-history",
  handler: (set: any, get: any) => ({
    history: [],
    setHistory: (event: any) => set({ history: get().history.concat(event) }),
    clearHistory: (event: any) => set({ history: [] }),
  }),
});

```

**Comment by kettanaito** [comment]
Almost all suggestions here still use `createJSONStorage`, which defeats the whole point of using IndexedDB. You _don't need_ to serialize your data. If you want to store serializable data, use other web storages, like `localStorage`. The point of IndexedDB is that it can store _non-serializable_ data: class instances, functions, etc. 

Otherwise, the suggestions are solid when it comes to implementing your own persist:

```ts
import type { PersistStorage } from 'zustand/middleware'
import { get, set, del as remove, type UseStore } from 'idb-keyval'

export function createIndexedDbStorage(
  store: UseStore,
): PersistStorage<any, Promise<void>> {
  return {
    async getItem(key) {
      return await get(key, store).then((value) => value || null)
    },
    async setItem(key, value) {
      await set(key, value, store)
    },
    async removeItem(key) {
      await remove(key, store)
    },
  }
}

```

```ts
export const store = createStore(
  persist(() => ({}), {
    name: 'this-will-be-used-as-a-field-name',
    version: 0.1,
    storage: createIndexedDbStorage(
      createIndexedDbStore('indexeddb-database-name', 'indexeddb-store-name'),
    ),
  }),
)
```

> [!IMPORTANT] Bear in mind that when using `idb-keyval` you **cannot have multiple stores within the same database**. You cannot do `createIndexedDbStore('db', 'one')` and `createIndexedDbStore('db', 'two')`. You either use the same db+store pair or opt out from `idb-keyval` as stated in their README.

**Reply by dai-shi** [reply]
Cool.

https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API

> IndexedDB lets you store and retrieve objects that are indexed with a key; any objects supported by the [structured clone algorithm](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Structured_clone_algorithm) can be stored.

Yeah, we don't need createJSONStorage. Would you like to open a PR to improve our docs?

**Reply by kettanaito** [reply]
Of course! What are the docs that need improvement?

I would also be happy to publish my tiny `createIndexedDbStorage()` utility to help others persist Zustand stores via IndexedDB. It relies on `idb-keyval` but allows reusing the same db+store combos for different stores, which `idb-keyval` intentionally doesn't support. Let me know if it would be beneficial to publish such a package.

**Reply by dai-shi** [reply]
> What are the docs that need improvement?

Maybe `./docs/integrations/persisting-store-data.md`. @dbritto-dev Any suggestions?



> I would also be happy to publish my tiny `createIndexedDbStorage()` utility to help others persist Zustand stores via IndexedDB.

Cool. Would you like to host it in https://github.com/zustandjs ? It's not necessary, and you can host it under your profile if you prefer.

**Reply by kettanaito** [reply]
Of course, it would be great to store it under Zustand! Thanks. 

I propose to name the function `createIndexedDBStorage` to be inline with the existing `createJSONStorage`. If you're okay with that, could you please create a repo and invite me to push the implementation?

**Reply by dai-shi** [reply]
Sent an invitation. Please feel free to create a new repository.

**Reply by kettanaito** [reply]
Thanks, @dai-shi. Created a repo and pushed the code: https://github.com/zustandjs/indexeddb-storage. Would be grateful for your review of everything there.

**Reply by kettanaito** [reply]
In fact, could you please setup up the GitHub and NPM tokens for publishing that package? It appears the org-wide `NPM_TOKEN` is a granular token and those have been discontinued and no longer allow publishing. 

The tokens are extremely easy to set up:

- Follow [this](https://github.com/ossjs/release?tab=readme-ov-file#generate-github-personal-access-token) for GitHub token.
- Create a publish NPM token with just enough permissions to publish the new package and set it as the repo's `NPM_TOKEN` secret.

Thanks!

**Reply by dai-shi** [reply]
- Created https://www.npmjs.com/package/zustand-indexeddb and sent an invitation.
- Could you rename the repo to https://github.com/zustandjs/zustand-indexeddb ?
- I configured [Trusted Publishers](https://docs.npmjs.com/trusted-publishers) for that repository. You should be able to check the configuration too.

**Reply by kettanaito** [reply]
Thanks, that's a much better approach! NPM_TOKEN shouldn't be needed with this. I will try publishing the package once I have a moment. 

Did you have the chance to review it? I have a few things I'd like to solve, especially type-wise. How do I correctly annotate a persist storage to notify the consumer that `setItem` and `getItem` became asynchronous? Zustand doesn't seem to infer that despite the storage being provided type arguments `<S, Promise<void>>`.

**Reply by dbritto-dev** [reply]
@kettanaito what about something like this, basically creating your own `PersistStorage` type helper:

```ts
import type { StorageValue } from 'zustand/middleware/persist'

export interface PersistStorage<S> {
  getItem: (
    name: string,
  ) => Promise<StorageValue<S> | null>
  setItem: (name: string, value: StorageValue<S>) => Promise<void>
  removeItem: (name: string) => Promise<void>
}
```

```ts
export function createIndexedDBStorage<S>(
  databaseName: string,
  storeName: string,
): PersistStorage<S> {
```

**Reply by kettanaito** [reply]
Isn't this what `PersistStorage` does out of the box? https://github.com/zustandjs/zustand-indexeddb/blob/ed65d124a2c37cf841df6938179c09f0df773513/src/index.ts#L19-L22

**Reply by dbritto-dev** [reply]
@kettanaito yes and no, `getItem` return type is `StorageValue<S> | null | Promise<StorageValue<S> | null>` instead of only `Promise<StorageValue<S> | null>`

**Comment by dai-shi** [comment]
> Did you have the chance to review it? I have a few things I'd like to solve, especially type-wise.

Hi, I looked at the code and it's pretty straightforward.
I personally don't love making `S` type annotation, as it's not safe. This also applies to `createJSONStorage`.

> Zustand doesn't seem to infer that despite the storage being provided type arguments `<S, Promise<void>>`.

If I understand it correctly, yeah, it doesn't distinguish sync storage and async storage.
It's the design choice a long time ago, but there should be a better way if I think of it now.

**Reply by kettanaito** [reply]
Got it, thanks!

npm errors have defeated me when I was trying to publish the package. I will retry it a bit later when I have a moment. Has to do with the granular tokens rollout, not sure what exactly is wrong as the npm output is 2-screens long with no indicative of what's actually causing the error.

**Comment by kettanaito** [comment]
## 2025 Update

If you're looking to persist your Zustand stores in IndexedDB, I've published a community package called [`zustand-indexeddb`](https://github.com/zustandjs/zustand-indexeddb). It's the most versatile way to use IndexedDB for store persistence and I hope it will serve your apps well. 

```sh
npm i zustand-indexeddb
```

Also check out the [FAQ](https://github.com/zustandjs/zustand-indexeddb?tab=readme-ov-file#faq) section to find answers to questions like "How is this different from using `idb-keyval`?" (because it is different, in a good way).

**Reply by m7md1alaa** [reply]
thank you so much

**Comment by wa1one** [comment]
There is hardcoded version of db version   `const openRequest = indexedDB.open(databaseName, 1)`, how is that gonna be used? One store per db?

**Reply by kettanaito** [reply]
That's likely an oversight and that version should be removed or the version should be surfaced to the user. Pull requests are welcome!
